### PR TITLE
Fix wide-angle lens flare occlusion lag

### LINKED ISSUE #3310: Incorrect wide angle camera lens flare occlusion after camera movement
## Environment
* Ubuntu 20.04
* gazebo11 d3584008b47c745c1732a6cd07af68726cd900fa

## Description
* Expected behavior: After moving a robot with a wide angle camera and then triggering an image, lens flare should be occluded or not occluded as expected.
* Actual behavior: Sometimes lens flare is not occluded by objects, and sometimes lens flare does not appear when there is no object to occlude it.

## Steps to reproduce
After moving your robot this call to `occlusionScale` will sometimes give an incorrect result: https://github.com/gazebosim/gazebo-classic/blob/gazebo11/gazebo/rendering/LensFlare.cc#L314

If you inspect the view matrix just before that call to `occlusionScale`, it will not be up to date. Move your robot, trigger an image, and then trigger a second image. You should see that the view matrix changes between image triggers. Here's some code for inspecting the view matrix:
```
            const Ogre::Matrix4& m = this->dataPtr->wideAngleDummyCamera->OgreCamera()->getViewMatrix();
            std::cerr << m[0][0] << "  " << m[0][1] << "  " << m[0][2] << "  " << m[0][3] << std::endl;
            std::cerr << m[1][0] << "  " << m[1][1] << "  " << m[1][2] << "  " << m[1][3] << std::endl;
            std::cerr << m[2][0] << "  " << m[2][1] << "  " << m[2][2] << "  " << m[2][3] << std::endl;
            std::cerr << m[3][0] << "  " << m[3][1] << "  " << m[3][2] << "  " << m[3][3] << std::endl;
```
Digging a little deeper, you can see that in `Scene::FirstContact` https://github.com/gazebosim/gazebo-classic/blob/gazebo11/gazebo/rendering/Scene.cc#L1217 the `origin` and `dir` used to compute occlusions are incorrect during the first image trigger because the view matrix is incorrect.

### PR DESCRIPTION
**Pull request by audrow**
Fixes #3310 

This forces an update of the wide angle dummy camera, which updates the Ogre camera's view matrix. 

---

One way to check that this fixes the bug, is to use [this](https://github.com/audrow/gazebo-classic/blob/audrow/debug-lens-flare/worlds/lensflare_wideangle_cam.world) world file. Open it in Gazebo, and then move the only camera (`wide_angle_cameras_occluded_higher`) from it's starting height (`z=21`) to `z=11`, where it will have a wall between it and the light source. You can then move it back to `z=21`.

You should see lens flare when the light source is in the image and now lens flare when the light source is occluded.

### CODE REVIEW

File: gazebo/rendering/LensFlare.cc [hunk 1]
```diff
@@ -305,6 +305,7 @@ namespace gazebo
                 ignition::math::Pose3d(
                   Conversions::ConvertIgn(cam->getDerivedPosition()),
                   Conversions::ConvertIgn(quat)));
+            this->dataPtr->wideAngleDummyCamera->OgreCamera()->getParentSceneNode()->_update(true, true);
```
Comment (iche033) [inline gazebo/rendering/LensFlare.cc]: nit: wrap line to 80 chars

Comment (audrow) [reply gazebo/rendering/LensFlare.cc]: Fixed in 1207519.

**Review by iche033 — APPROVED** [review]

**Comment by scpeters** [comment]
this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this

**Comment by scpeters** [comment]
> this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this

I've gotten a test working in branch `scpeters/lens-flare-occlusion-lag`. I opened a pull request targeting this PR in https://github.com/audrow/gazebo-classic/pull/3, but it didn't start Jenkins jobs because it's targeting your fork. If the test looks ok, then please merge it into this PR with a merge commit and we can confirm that the tests pass in CI

**Comment by scpeters** [comment]
> > this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this
> 
> I've gotten a test working in branch `scpeters/lens-flare-occlusion-lag`. I opened a pull request targeting this PR in [audrow#3](https://github.com/audrow/gazebo-classic/pull/3), but it didn't start Jenkins jobs because it's targeting your fork. If the test looks ok, then please merge it into this PR with a merge commit and we can confirm that the tests pass in CI

the new test case passed on Ubuntu but failed in the homebrew build on [mac-four.bigsur](https://build.osrfoundation.org/computer/mac-four.bigsur). I noticed all the GUI tests failed on that run as well, so there may be an issue with the CI machine. I just pushed an improvement to the image waiting logic in https://github.com/gazebosim/gazebo-classic/pull/3325/commits/29dbe5a03994dd7fb17b541ba98a4caf9494c173 that hopefully is more robust. The new CI build is also running on [mac-four.bigsur](https://build.osrfoundation.org/computer/mac-four.bigsur), so we'll see if it helps

**Comment by scpeters** [comment]
now with https://github.com/gazebosim/gazebo-classic/commit/29dbe5a03994dd7fb17b541ba98a4caf9494c173 it passed on homebrew but failed repeatedly on Ubuntu. I'll keep testing locally to see if I can reproduce the flakiness

**Comment by scpeters** [comment]
> now with [29dbe5a](https://github.com/gazebosim/gazebo-classic/commit/29dbe5a03994dd7fb17b541ba98a4caf9494c173) it passed on homebrew but failed repeatedly on Ubuntu. I'll keep testing locally to see if I can reproduce the flakiness

I was able to reproduce the issue locally, and it appears in my local testing that `--lockstep` ( https://github.com/gazebosim/gazebo-classic/pull/3325/commits/032247ef245b984d4bc4b9230b2382edc0028b4c ) fixes the sensor timing and test flakiness

**Review by scpeters — APPROVED** [review]

**Comment by scpeters** [comment]
test passes in both CI runs; merging

**Pull request merged**